text/0000-inferred-type-aliases.md

```diff
@@ -0,0 +1,696 @@
+`_` inside `type Alias = _;` is the [transparent] version of the [opaque]
+and [inferred] `type Alias = impl Trait` ([RFC 2515]) or equivalently
+`existential type Alias: Trait;` ([RFC 2071]).
+
+## Dos and Don’ts
+[do_dont]: #dos-and-don'ts
+
+1. **Do not** use `_` inside type aliases as a means of *abstraction*.
+  Aliases that use `_` inside them will *leak* all information about the
+  exact type that is inferred. Changes to the capabilities / affordances /
+  operations offered by the underlying inferred type will cause breakage in
+  consumers of your APIs that depend on those affordances. Therefore, be extra
+  careful when using `pub` on `type Alias = _;` or when using `_` for an
+  associated type of an implementation for a public trait + public type.
+
+2. **Do** use `_` as a means of eliding complex implementation details that
+  are not important to the users of your API.
+  This applies in particular to custom derive macros for traits with
+  associated types.
+
+3. **Do**, assuming that you keep in mind 1), use `_` as a means to name the
+   type of an expression with a [voldemort type] somewhere inside it.
+
+## Teaching this
+
+The teaching of the [transparent] and [inferred] types can be explained in
+conjunction with explaining the usual `_` feature in other places.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar
+
+The grammar of `type` aliases and associated `type`s are extended to permit
+an optional bound (*"trait ascription"* if you will).
+To do this, we replace the production:
+```rust
+impl_type : attrs_and_vis maybe_default item_type ;
+
+item_type
+: TYPE ident generic_params maybe_where_clause '=' ty_sum ';' ;
+```
+
+with:
+
+```rust
+impl_type : attrs_and_vis maybe_default item_type ;
+
+item_type
+: TYPE ident generic_params maybe_bounds maybe_where_clause '=' ty_sum ';' ;
+```
+
+## Type checking
+
+Currently, if a user writes (associated type or type alias does not matter):
+
+```rust
+type Foo = _;
+```
+
+[`E0121`]: https://doc.rust-lang.org/stable/error-index.html#E0121
+
+they are presented with the error [`E0121`]:
+
+```rust
+error[E0121]: the type placeholder `_` is not allowed within types on item signatures
+  --> src/main.rs:11:12
+   |
+11 | type Foo = _;
+   |            ^ not allowed in type signatures
+```
+
+This restriction is lifted by this RFC in `type` items.
+
+Instead, the `_` "type" is permitted within `type`s.
+However, each `type` item that uses `_` in any part of its definition
+must be constrained by at least one `fn` body, or `const`/`static` initializer
+(henceforth: *"determinant"*). All of these determinants must be in the same
+module as the `type` item inferred and must independently place constraints
+on the `type` item such that they determine the same underlying exact type for it.
+A body or initializer must either fully constrain or place no constraints upon a given `type` item with `_` inside it.
```

Comment:
This was specifically chosen to be the case for `existential type`. Are we sure it makes sense here too? It would be a breaking change to alter this later.

Reply:
For consistency it seems like the right thing to do and to also not introduce full global type inference...

Elaborate on why it would be a breaking change in the future?

Reply:
I'm making mistakes all over the place this morning. Laxening the inference does seem forward-compatible. It seems consistent with `existential trait`, but not with `_` in other contexts.

Reply:
Added a paragraph on ~this in the rationale.